esx_service/vmdk_ops.py

```diff
@@ -209,6 +213,87 @@ def make_create_cmd(opts, vmdk_path):
         return "{0} -d {1} -c {2} {3}".format(VMDK_CREATE_CMD, disk_format, size, vmdk_path)
 
 
+def cloneVMDK(vm_name, vmdk_path, src_vmdk_path, opts={}, vm_uuid=None, vm_datastore=None):
+    logging.info("*** cloneVMDK: %s opts = %s", vmdk_path, opts)
+
+    # Get source volume path for cloning
+    error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
+    if error_info:
+        return err(error_info)
+
+    try:
+        src_volume, src_datastore = parse_vol_name(opts["clone-from"])
+    except ValidationError as ex:
+        return err(str(ex))
+    if not src_datastore:
+        src_datastore = vm_datastore
+    elif src_datastore not in known_datastores():
+        return err("Invalid datastore '%s'.\n" \
+                    "Known datastores: %s.\n" \
+                    "Default datastore: %s" \
+                    % (datastore, ", ".join(known_datastores()), vm_datastore))
+    src_path = get_vol_path(src_datastore, tenant_name)
+    if src_path is None:
+        return err("Failed to initialize source volume path {0}".format(src_path))
+    src_vmdk_path = vmdk_utils.get_vmdk_path(src_path, src_volume)
+
+    # We need to reauthorize with size info of the volume being cloned
+    if vm_uuid and vm_datastore:
+        src_vol_info = kv.get_vol_info(src_vmdk_path)
+        opts["size"] = src_vol_info["size"]
+        error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, vm_datastore, "create", opts)
+        if error_info:
+            return err(error_info)
+
+    attached, uuid, attach_as = getStatusAttached(src_vmdk_path)
+    if attached:
+        return err("Source volume cannot be in use when cloning")
+
+    if not kv.DISK_ALLOCATION_FORMAT in opts:
+        disk_format = kv.DEFAULT_ALLOCATION_FORMAT
+    else:
+        disk_format = str(opts[kv.DISK_ALLOCATION_FORMAT])
+
+    # VirtualDiskSpec
+    vdisk_spec = vim.VirtualDiskManager.VirtualDiskSpec()
+    vdisk_spec.adapterType = 'busLogic'
+    vdisk_spec.diskType = disk_format
+
+    # Form datastore path from vmdk_path
+    dest_vol = vmdk_utils.get_datastore_path(vmdk_path)
+    source_vol = vmdk_utils.get_datastore_path(src_vmdk_path)
+
+    si = get_si()
+    task = si.content.virtualDiskManager.CopyVirtualDisk(
+        sourceName=source_vol, destName=dest_vol, destSpec=vdisk_spec)
```

Comment:
Changing to API calls on the *VMDK ops would be a straightforward change, if I manage to get time I can investigate it further